utils/zerovec/src/ule/macros.rs

```diff
@@ -0,0 +1,21 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+#[macro_export]
+macro_rules! impl_ule_from_array {
+    ($source:ty, $dest:ty, $single:path, $zero:expr) => {
+        /// Convert an array of `$source` to an array of `$dest`.
+        pub const fn from_array<const N: usize>(arr: [$source; N]) -> [Self; N] {
+            let mut result = [$zero; N];
```

Comment:
You could get around the need for `$zero` by initialising the array to the first value. You'd need a special case for `N == 0`, but that will optimize out.

Reply:
Offline discussion with @robertbastian:
The following `N == 0` special cases didn't work:
* `core::mem::transmute(())` - the compiler couldn't infer that `N` is `0`
* `core::mem::transmute(arr)` - the compiler couldn't infer that `N` is `0`
* `[]` - the compiler couldn't infer that `N` is `0`
* `[core::mem::zeroed(); N]` - `zeroed` is not const

I am using `*(&arr as *const _ as *const [Self; N])` now, which compiles, but I am not sure if that's UB. Could someone weigh in?

Reply:
I believe this is fine, as it's a zero-sized pointer read from valid memory. Actually I found this though:

> The canonical way to obtain a pointer that is valid for zero-sized accesses is [NonNull::dangling](https://doc.rust-lang.org/std/ptr/struct.NonNull.html#method.dangling)

https://doc.rust-lang.org/std/ptr/index.html

Reply:
That's perfect, thanks!